### ISSUE TITLE
Memory inefficiency in RNNs

### ISSUE THREAD

**Issue by atmguille** [keras-team/keras#16113]
**Context**
I have been working with small `ConvLSTM` architecture and getting an OOM with a moderate number of filters. I think I have detected a memory inefficiency related to RNNs that may be the root cause, although I have not been able to fully solve the problem.

**System information**.
TensorFlow version (you are using): 2.6.2
GPU model and memory: Tesla P100 16GB (Kaggle GPU)
Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state**.
At the moment, `keras.backend.rnn` returns the tuple `(last_output, outputs, new_states)`. Layers that call this method may only require *last_output*, so retrieving and storing *outputs* in memory, in this case, is inefficient and not needed.

The current behavior is especially harmful to Convolutional versions of RNNs and LSTMs, since storing *outputs* means storing something close to n_filters * input. For example, if a `ConvLSTM` with 50 filters 3x3 is instantiated and is applied to a video with 100x100 frames, `keras.backend.rnn` will return 50 copies of the complete video (98x98 frames) even if `return_sequences=False`, when just the last 98x98 frame for each filter is needed. 

As it can be seen from the example, if the current state is as described, the memory impact can be huge, restricting the use of ConvLSTMs to a moderate number of filters even with simple architectures.

**Will this change the current api? How?**
Yes. `keras.backend.rnn` will receive a flag (*return_all_outputs*) indicating if all outputs from the process have to be returned, and thus kept in memory. If not, only last_output will be kept during the process, saving the corresponding memory.
To keep the current return format, if *return_all_outputs* is False, the method can still return the tuple `(last_output, [last_output], new_states)`.

**Who will benefit from this feature?**
The whole community, especially users using Convolutional versions of RNNs and LSTMs. Other users have reported similar memory issues with ConvLSTMs in StackOverflow (for instance: [link](https://stackoverflow.com/questions/66859089/getting-oom-with-small-convlstm-model)).

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): yes
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing): the solution described in the API Changes has been implemented in [this repo](https://github.com/atmguille/convlstm-saving-memory). Some TODOs are pending, like applying masks correctly to outputs when *return_all_outputs* is False. However, other updates are needed, since it does not solve the OOM of my original problem.

**Standalone code to reproduce the issue**.
Check [this notebook](https://www.kaggle.com/atmguille/convlstm-oom/notebook#ConvLSTM) to see how a simple architecture results in an OOM. Especially worrying is the following line of the provided logs: 
```
2022-02-23 00:08:53.579673: I tensorflow/core/common_runtime/bfc_allocator.cc:1011] Bin (16777216): 	Total Chunks: 826, Chunks in use: 826. 14.79GiB allocated for chunks. 14.79GiB in use in bin. 14.78GiB client-requested in use in bin.
```
Check [this notebook](https://www.kaggle.com/atmguille/convlstm-oom-using-proposal/notebook#Load-proposal-code) that imports my proposal, which also gets an OOM.

**Comment by mattdangerw** [comment]
@atmguille thanks for filing! Would you be willing to contribute a PR here?

**Comment by atmguille** [comment]
@mattdangerw sure! I have already a proposal in [the repository](https://github.com/atmguille/convlstm-saving-memory) I mention in the issue. I have done some experiments in CPU and experienced better performance in RAM usage for small number of filters. However, as I mentioned in the issue, I may be missing something else for GPU.

Thanks for your time. Let me know what are the next steps!

**Comment by mattdangerw** [comment]
Tagging @qlzh727 and @fchollet who probably know the most about Keras RNN layers to see if they have thoughts.

I would say next could be opening a pull request and we can continue discussion there.

It would be good to verify this functions with a GPU, could you try running your proposed changes with a GPU runtime in a colab?

**Comment by atmguille** [comment]
@mattdangerw the code works in GPU. I have tested it in Kaggle, [the notebook](https://www.kaggle.com/atmguille/convlstm-oom-using-proposal#ConvLSTM) that uses my proposal is able to fit the model in memory while [the notebook](https://www.kaggle.com/atmguille/convlstm-oom#ConvLSTM) that uses the current state of Keras is not. 

I agree with you, we can discuss further details of the code in the PR. I will try to open it as soon as I can. 

Thanks again for your time and dedication!
